### PR TITLE
feat(scim): idp flag ui permissions changes

### PR DESCRIPTION
**Pull request by cathteng**
Cannot remove a member from the org if they are `idp:provisioned` and cannot leave the org if you are `idp:provisioned`
https://user-images.githubusercontent.com/70817427/210852591-4abf1b69-15a0-402b-9a5e-135f894a285d.mov

Cannot change your or a member's role if you/they are `idp:role-restricted`
https://user-images.githubusercontent.com/70817427/210882081-1d67c719-e4d3-4d7c-81cb-989cd97c3dd2.mov

Cannot join or leave a team if the team is `idpProvisioned` (the same component is used for adding a team to a project -- you *can* add a team to a project even if the team is `idpProvisioned`
https://user-images.githubusercontent.com/70817427/210882015-3cae148c-9f50-4af7-882d-af96afff2c55.mov

Addresses ER-1289

[ER-1289]: https://getsentry.atlassian.net/browse/ER-1289?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: static/app/views/settings/components/teamSelect.tsx
Comment (AniketDas-Tekky) [reply static/app/views/settings/components/teamSelect.tsx:169]: Something like "Team membership is controlled by an identity provider. Contact your administrator to be removed from this team"?

**Review by AniketDas-Tekky — COMMENTED** [review]
Besides wording, it looks pretty good. Don't merge this though until I get the feature flag checked in for idp restriction.

File: static/app/views/settings/components/teamSelect.tsx
Comment (cathyteng17) [reply static/app/views/settings/components/teamSelect.tsx:95]: idp provisioned teams can only have users added to them through the identity provider

File: static/app/types/organization.tsx
Comment (leedongwei) [inline static/app/types/organization.tsx:80]: Would it make sense if this property is nested inside a `flags` key, similar to `Member`?

File: static/app/views/settings/components/teamSelect.tsx
Comment (leedongwei) [reply static/app/views/settings/components/teamSelect.tsx:95]: @cathyteng17 can you check that this component is only used for assigning `Members` into `Teams`. 

It might make sense to add a comment or change the name of the component to make it explicitly clear that this is cannot be used for generic assignment of "X" to `Teams`, because of the IdP-provisioned restriction.

File: static/app/views/settings/components/teamSelect.tsx
Comment (leedongwei) [inline static/app/views/settings/components/teamSelect.tsx:100]: I don't think we should filter out IdP-provisioned in this dropdown. 

As a user, it would be frustrating that I cannot find a team in this dropdown and I wouldn't know that it was not included because it was IdP-provisioned. IMO they should be included, but disabled and labeled differently.

File: static/app/views/settings/components/teamSelect.tsx
Comment (leedongwei) [reply static/app/views/settings/components/teamSelect.tsx:169]: > Membership to this team is managed through your organization's identity provider.

File: static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx
Comment (leedongwei) [reply static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx:44]: > This member's organization-level role is managed through your organization's identity provider.

Replace "this member's" with "your" if the user is looking at their own OrganizationMember page.

File: static/app/views/settings/organizationMembers/organizationMemberRow.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationMembers/organizationMemberRow.tsx:216]: It makes the code rather brittle to be checking for the IdP flag at this place. 

You should rework the logic.

File: static/app/views/settings/organizationMembers/organizationMemberRow.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationMembers/organizationMemberRow.tsx:235]: Reuse the "_XYZ_ is managed through your organization's identity provider." template.

File: static/app/views/settings/organizationTeams/allTeamsRow.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationTeams/allTeamsRow.tsx:220]: Reuse the template.

File: static/app/views/settings/organizationTeams/teamMembers.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationTeams/teamMembers.tsx:82]: You shouldn't need to fetch this yourself because `Team` is a commonly used model on the frontend, so there should be a cache on the frontend that you can use. Look for the file `teamStore.tsx`.

**Review by leedongwei — CHANGES_REQUESTED** [review]
Overall it looks good, just some things that needs to be addressed before merge.

File: static/app/types/organization.tsx
Comment (cathyteng17) [reply static/app/types/organization.tsx:80]: this is currently its own column in the Team model, so i have it as its own field here https://github.com/getsentry/sentry/blob/bc9b2d8ef60da18a3f775428b0743b499e8faa45/src/sentry/models/team.py#L156

File: static/app/views/settings/components/teamSelect.tsx
Comment (cathyteng17) [reply static/app/views/settings/components/teamSelect.tsx:95]: this component is also used to assign a project to a team... https://github.com/getsentry/sentry/blob/0dbe496f9d856824206072b2a192b61c1ca8f5da/static/app/views/settings/project/projectTeams.tsx#L149

File: static/app/views/settings/components/teamSelect.tsx
Comment (cathyteng17) [reply static/app/views/settings/components/teamSelect.tsx]: i made them grayscale and with a tooltip with the template text if they're idpProvisioned

File: static/app/views/settings/components/teamSelect.tsx
Comment (cathyteng17) [reply static/app/views/settings/components/teamSelect.tsx]: i added an extra prop to make sure adding an idp-provisioned team is only disabled if this component is being used for org members adding teams

File: static/app/views/settings/organizationTeams/teamMembers.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationTeams/teamMembers.tsx:76]: Can you check if this is actually needed? 

It seems like `this.props.team` is populated for this AsyncView because it is passed in by `<TeamDetails/>`, so it might make sense to use that instead? You'll need to update the prop typing.

**Review by leedongwei — CHANGES_REQUESTED** [review]
Looks good! Most stuff are in place. Just small changes needed.

File: static/app/views/settings/organizationTeams/allTeamsRow.tsx
Comment (cathyteng17) [reply static/app/views/settings/organizationTeams/allTeamsRow.tsx:259]: there are buttons with different behaviors upon clicking depending on whether or not the team has open membership

Comment (cathyteng17) [reply static/app/views/settings/organizationTeams/allTeamsRow.tsx:259]: i can make a helper function

File: static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx
Comment (cathyteng17) [reply static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx:52]: Screen Shot 2023-01-05 at 09 29 21
we settled on this

File: static/app/views/settings/organizationTeams/teamMembers.tsx
Comment (cathyteng17) [reply static/app/views/settings/organizationTeams/teamMembers.tsx:76]: it looks like none of the fetching the team from the TeamStore stuff is necessary since the team is already a prop

File: static/app/types/organization.tsx
Comment (cathyteng17) [reply static/app/types/organization.tsx:80]: https://github.com/getsentry/sentry/pull/42166#pullrequestreview-1210627208

File: static/app/views/settings/components/teamSelect.tsx
Comment (leedongwei) [inline static/app/views/settings/components/teamSelect.tsx:28]: For boolean props, it's generally good practice to start it with `is` or `has`. 

However, there's precedence on the `OrganizationMemberDetail` page where the `OrganizationRoleSelect` is using `enforceAllowed` and `enforceRetired` (see [link](https://github.com/getsentry/sentry/blob/2eac063cf8989dfc0791ca8b12c78575631b22ac/static/app/views/settings/organizationMembers/organizationMemberDetail.tsx#L356)). 

As such, a better name for this property would be `enforceIdpProvisioned`, then on the Teams Page, set it to `false`.

File: static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx:51]: Might be better to do it further below, in this manner 

```suggestion
      tct("[person] organization-level role is managed through your organization's identity provider.", {person: isCurrentUser ? "Your" : "This member's"});
```

File: static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx:27]: Might be better to name it as `enforceIdpRestricted`.

File: static/app/views/settings/components/teamSelect.tsx
Comment (leedongwei) [inline static/app/views/settings/components/teamSelect.tsx:113]: Need to wrap this in `t('...')`

File: static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationMembers/inviteMember/orgRoleSelect.tsx:50]:
```suggestion
          <div>{t('Organization Role')}</div>
```

File: static/app/views/settings/organizationMembers/organizationMemberDetail.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationMembers/organizationMemberDetail.tsx:362]:
```suggestion
          enforceAllowed={false}
          enforceIdpRoleRestricted={member.flags['idp:role-restricted']}
          enforceRetired={hasTeamRoles}
          isCurrentUser={isCurrentUser}
```

File: static/app/views/settings/organizationTeams/teamMembers.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationTeams/teamMembers.tsx:267]: Nit: I don't think you need the `?` because `team` is guaranteed to be there.

Comment (cathyteng17) [reply static/app/views/settings/organizationTeams/teamMembers.tsx:267]: teamMembers.spec.jsx fails if the ? isn't there

Comment (cathyteng17) [reply static/app/views/settings/organizationTeams/teamMembers.tsx:267]: i think the `team` loads slower than the component loads, so for a little bit, team is `undefined` and the test fails if there's no `?`

File: static/app/views/settings/organizationTeams/teamMembers.spec.jsx
Comment (leedongwei) [inline static/app/views/settings/organizationTeams/teamMembers.spec.jsx:24]: Might make more sense to update the stub with 

```
    flags: {
      'idp:provisioned': false,
    },
```

so it can be consistent across all tests.

This can be a separate PR.

File: static/app/views/settings/organizationMembers/organizationMemberDetail.tsx
Comment (leedongwei) [inline static/app/views/settings/organizationMembers/organizationMemberDetail.tsx:378]: Nit: The order of props should be consistent to make it easier to read. `enforceXYZ` came first for `OrganizationRoleSelect`, it should be the same for `TeamSelect` too.

File: static/app/types/organization.tsx
Comment (leedongwei) [reply static/app/types/organization.tsx]: "Fixed" with https://github.com/getsentry/sentry/pull/42851

**Review by leedongwei — APPROVED** [review]
LGTM! 🎉

**Pull request merged**
